pkg/storage/couchdb/couchdbstore.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package couchdbstore
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+
+	_ "github.com/go-kivik/couchdb" // The CouchDB driver
+	"github.com/go-kivik/kivik"
+
+	"github.com/trustbloc/edge-core/pkg/storage"
+)
+
+// Provider represents an CouchDB implementation of the storage.Provider interface
+type Provider struct {
+	hostURL       string
+	couchDBClient *kivik.Client
+	dbs           map[string]*CouchDBStore
+}
+
+const (
+	blankHostErrMsg           = "hostURL for new CouchDB provider can't be blank"
+	failToCloseProviderErrMsg = "failed to close provider"
+)
+
+// NewProvider instantiates Provider
+func NewProvider(hostURL string) (*Provider, error) {
+	if hostURL == "" {
+		return nil, errors.New(blankHostErrMsg)
+	}
+
+	client, err := kivik.New("couch", hostURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Provider{hostURL: hostURL, couchDBClient: client, dbs: map[string]*CouchDBStore{}}, nil
+}
+
+// CreateStore creates a new store with the given name.
+func (p *Provider) CreateStore(name string) error {
+	err := p.couchDBClient.CreateDB(context.Background(), name)
+	return err
+}
+
+// OpenStore opens an existing store with the given name and returns it.
+func (p *Provider) OpenStore(name string) (storage.Store, error) {
+	exists, err := p.storeExists(name)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, storage.ErrStoreNotFound
+	}
+
+	db := p.couchDBClient.DB(context.Background(), name)
+
+	if dbErr := db.Err(); dbErr != nil {
+		return nil, dbErr
+	}
+
+	store := &CouchDBStore{db: db}
+
+	p.dbs[name] = store
```

Comment:
You need a write lock here.

Reply:
After looking through the OpenStore method, I think that I actually need a write lock for the whole method since it first checks to see if it exists, and then opens the store in the map. Let me know if you think I'm locking more than I need to here.